File: java-class-loader/vm.java
//package

/*

The "VM" - a front end to the classloader.
Ported from C++ to Java.

Copyright 2015 Sam Saint-Pettersen

Released under the MIT/X11 License.
Please see LICENSE file.

*/

class VM {

	private static String version = "1.0";

	private static void displayVersion() {

		String javaVersion = System.getProperty("java.version");
		String javaVendor = System.getProperty("java.vendor");
		String versionInfo = String.format("Java \"VM\" v. %s (running on Java %s by %s).", version, javaVersion, javaVendor);
		System.out.println(versionInfo);
		System.exit(0);
	}

	private static void loadClassFile(String claSS, boolean dump) {

		ClassLoader classLoader = new ClassLoader();
		classLoader.load(String.format("%s.class", claSS), dump);
	}

	public static void main(String[] args) {

		if(args.length == 0) System.out.println("Usage: VM -c [-d] <classFile>");

		else if(args.length > 0 && args.length < 7) {

			for(int i = 0; i < args.length; ++i) {

				if(args[i].equals("-c") || args[i].equals("--class")) {
					if(args[i+1].equals("-d") || args[i+1].equals("--dump")) {

						loadClassFile(args[i+2], true);
					}
					else {

						loadClassFile(args[i+1], false);
					}
				}
				else if(args[i].equals("-v") || args[i].equals("--version")) {

					displayVersion();
				}
			}
		}
	} 
}

File: lua-class-loader/vm.lua
-- The "VM" - a front end to the ClassLoader.
-- Ported from C++ to Lua.
--
-- Copyright 2015 Sam Saint-Pettersen
--
-- Released under the MIT/X11 License.
-- Please see LICENSE file.

require 'classloader'

VM_VERSION = '1.0'

function displayVersion()
	print(string.format('Java \"VM\" v. %s', VM_VERSION))
	os.exit(0)
end

function loadClassFile(class, dump)
	local classLoader = ClassLoader.create()
	classLoader:load(string.format('%s.class', class), dump)
	os.exit(0)
end

function main()
	if #arg == 0 then 
		print('Usage: vm -c [-d] <classFile>\n') 
	else
		if arg[1] == '-v' or arg[1] == '--version' then
			displayVersion()
		end
		local i = 1
		while i < #arg do
			if arg[i] == '-c' or arg[i] == '--class' then
				if arg[i+1] == '-d' or arg[i+1] == '--dump' then
					loadClassFile(arg[i+2], true)
				else
					loadClassFile(arg[i+1], false)
				end
			end
		end
	end
	os.exit(0)
end

main()
